trellis-triplestore/src/main/java/org/trellisldp/triplestore/TriplestoreResourceService.java

```diff
@@ -141,58 +140,70 @@ private static RDFTerm getBaseIRI(final RDFTerm object) {
-                getContainer(identifier).ifPresent(parent ->
-                        dataset.add(PreferServerManaged, identifier, DC.isPartOf, parent));
+    @Override
+    public Future<Boolean> replace(final IRI identifier, final IRI ixnModel, final Dataset dataset) {
+        LOGGER.debug("Updating: {}", identifier);
+        return supplyAsync(() -> createOrReplace(identifier, ixnModel, dataset, OperationType.REPLACE));
+    }
 
-            }
-            final Literal time = rdf.createLiteral(eventTime.toString(), XSD.dateTime);
-
-            try {
-                rdfConnection.update(buildUpdateRequest(identifier, time, dataset));
-                emitEvents(identifier, time, dataset);
-                if (!isDelete) {
-                    mementoService.ifPresent(svc -> get(identifier).ifPresent(res ->
-                                svc.put(identifier, eventTime, res.stream())));
-                }
+    private Boolean createOrReplace(final IRI identifier, final IRI ixnModel, final Dataset dataset,
+            final OperationType type) {
+        final Instant eventTime = now();
+
+        // Set the LDP type
+        dataset.remove(of(PreferServerManaged), identifier, RDF.type, null);
+        dataset.add(PreferServerManaged, identifier, RDF.type, ixnModel);
+
+        // Relocate some user-managed triples into the server-managed graph
+        if (LDP.DirectContainer.equals(ixnModel) || LDP.IndirectContainer.equals(ixnModel)) {
+            dataset.getGraph(PreferUserManaged).ifPresent(g -> {
+                g.stream(identifier, LDP.membershipResource, null).findFirst().ifPresent(t -> {
+                    // This allows for HTTP resource URL-based queries
+                    dataset.add(PreferServerManaged, identifier, LDP.member, getBaseIRI(t.getObject()));
+                    dataset.add(PreferServerManaged, identifier, LDP.membershipResource, t.getObject());
+                });
+                g.stream(identifier, LDP.hasMemberRelation, null).findFirst().ifPresent(t -> dataset
+                                .add(PreferServerManaged, identifier, LDP.hasMemberRelation, t.getObject()));
+                g.stream(identifier, LDP.isMemberOfRelation, null).findFirst().ifPresent(t -> dataset
+                                .add(PreferServerManaged, identifier, LDP.isMemberOfRelation, t.getObject()));
+                dataset.add(PreferServerManaged, identifier, LDP.insertedContentRelation,
+                                g.stream(identifier, LDP.insertedContentRelation, null).map(Triple::getObject)
+                                                .findFirst().orElse(LDP.MemberSubject));
+            });
+        }
+        // Set the parent relationship
+        getContainer(identifier).ifPresent(parent -> dataset.add(PreferServerManaged, identifier, DC.isPartOf, parent));
+        mementoService.ifPresent(svc -> get(identifier).ifPresent(res -> svc.put(identifier, eventTime, res.stream())));
```

Comment:
I've written some Memento tests, and I've run into an issue related to this line that will require a little thought.

Imagine this scenario:

1. An LDP-RS is created with some number of triples.
2. Some time passes, and the triples are changed.
3. Some more time passes and some of the triples are changed again.

Now, a simple GET request will fetch the current state of the resource, but how many Memento resources are there?

Given the code here in your branch: https://git.io/vAaMW there will be two Mementos.

Based on the code currently in master: https://git.io/vAaSv there will be three Mementos.

The difference, code-wise, has to do with when the `::get(identifier)` is called when passing data to the mementoService. In your version, that is called before the new data is stored, in the current master, it is called after the data is stored.

In another sense, the answer depends on whether a memento should be created immediately when a resource is created (that is, the latest Memento is always equivalent to the current state of the resource) or whether creation should happen only after the current state changes. My main consideration for the current behavior has to do with client expectations. That is, if the state of a resource at time X is X' and the state of the resource at time Y is Y', then a request for the resource at (X <= time < Y) should return resource state X'. Likewise, a request for the resource at (Y <= time) should return Y'. But the code in your PR offsets the Mementos. That is, if the resource is changed exactly at time Y, then a request for a Memento at time Y will return X'. Assume that the time precision is 1 second, due to HTTP spec limitations.

Reply:
Another thought on this and the issue above about the memento service -- both are related specifically to the triplestore impl and somewhat orthogonal to the main thrust of this PR. It's up to you, but it would be totally find to leave the implementation code as-is and deal with it separately. I certainly don't want to hold up this effort.